### PR TITLE
[GH-235] Refresh token on user level oauth (#235)

### PR DESCRIPTION
**Pull request by jupriano**
#### Summary

i think is only move the refresh token outside "if isAccountLevel" but if the refresh token function only for account level, how we get user level token ?
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-zoom/issues/235

### CODE REVIEW

**Comment by larkox** [comment]
Thanks, @vicky-demansol but this won't work. If I understand the code correctly, you are just not making any distinction between AccountLevel and UserLevel by removing the if. Both tokens come from different places (take a look to the `getActiveClient` function), and have to be handled slightly different.

This may work if you instead of getting the token from the `getSuperUserToken` you get it directly from the client (since it is already fetch from the database), but then you also have to distinguish where you store it. You must store the account level one as the superuser token, or the user level one on the user information.

File: server/http.go [hunk 1]
```diff
@@ -126,6 +126,12 @@ func (p *Plugin) completeUserOAuthToZoom(w http.ResponseWriter, r *http.Request)
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 			return
 		}
+	} else {
+		err = p.setUserToken(userID, token)
```
Comment (larkox) [inline server/http.go:130]: This seems cleaner so far, but a few lines after (line 161) we are storing the token along the OAuthUserInfo. If we want to follow this pattern, we would need to refactor the token out of the OAuthUserInfo and fix whatever breaks from that.

Another solution would be to keep the current pattern, and just update the token when needed in the OAuthUserInfo structure when needed.

I leave up to you which path to take.

File: server/plugin.go [hunk 2]
```diff
@@ -175,7 +175,10 @@ func (p *Plugin) getActiveClient(user *model.User) (Client, string, error) {
 	if err != nil {
 		return nil, message, errors.New("could not decrypt OAuth access token")
 	}
-
+	oldToken, err := p.getUserToken(user.Id)
+	if err == nil {
```
Comment (larkox) [inline server/plugin.go:179]: We should handle the error, and remove this (I guess) debug lines.

**Review by larkox — CHANGES_REQUESTED** [review]
I like the approach, but I fear it may not be handling all the cases correctly. Take a look at the comments.

File: server/plugin.go (hunk 2 above)
Comment (sibasankarnayak) [inline server/plugin.go:179]: if err != nil {
        p.API.LogInfo("error while fetching user token", "error",err.Error())
        retrun err
}

File: server/plugin.go [hunk 3]
```diff
@@ -175,7 +175,10 @@ func (p *Plugin) getActiveClient(user *model.User) (Client, string, error) {
...
+		p.API.LogInfo(fmt.Sprintf("oldTokenIs===========================%s", *oldToken))
+	}
```
Comment (sibasankarnayak) [inline server/plugin.go:181]:
```suggestion
	if err == nil {
		p.API.LogInfo(fmt.Sprintf("oldTokenIs===========================%s", *oldToken))
	}
```

File: server/plugin.go [hunk 4]
```diff
@@ -246,3 +249,22 @@ func (p *Plugin) SetZoomSuperUserToken(token *oauth2.Token) error {
 	}
 	return nil
 }
+
+func (p *Plugin) GetZoomUserToken(userID string) (*oauth2.Token, error) {
+	token, err := p.getUserToken(userID)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get token")
+	}
```
Comment (sibasankarnayak) [inline server/plugin.go:257]:
```suggestion
	}

```

File: server/plugin.go [hunk 5]
```diff
@@ -246,3 +249,22 @@ func (p *Plugin) SetZoomSuperUserToken(token *oauth2.Token) error {
...
+	if token == nil {
+		return nil, errors.New("zoom app not connected")
+	}
+	return token, nil
```
Comment (sibasankarnayak) [inline server/plugin.go:261]:
```suggestion

	return token, nil
```

File: server/plugin.go [hunk 6]
```diff
@@ -246,3 +249,22 @@ func (p *Plugin) SetZoomSuperUserToken(token *oauth2.Token) error {
...
+}
+
+func (p *Plugin) SetZoomUserToken(userID string, token *oauth2.Token) error {
+	err := p.setUserToken(userID, token)
+	if err != nil {
+		return errors.Wrap(err, "could not set token")
```
Comment (sibasankarnayak) [inline server/plugin.go:267]: error should be logged here or in the function getting called

File: server/zoom/oauth.go [hunk 7]
```diff
@@ -97,6 +97,7 @@ func (c *OAuthClient) GetMeeting(meetingID int) (*Meeting, error) {
 
 func (c *OAuthClient) getUserViaOAuth(user *model.User) (*User, error) {
 	url := fmt.Sprintf("%s/users/me", c.apiURL)
+
 	if c.isAccountLevel {
```
Comment (sibasankarnayak) [inline server/zoom/oauth.go:101]: remove unwanted change here

**Review by sibasankarnayak — APPROVED** [review]
Some non blocking changes are pending , code looks good so far.

**Review by larkox — APPROVED** [review]
LGTM, Thanks!

**Comment by dipak-demansol** [comment]
@hanzei can you pls tell me what i have to Test in this PR? what is the expected result for this?

**Comment by dipak-demansol** [comment]
> @larkox Can you please help with tests steps here? I'm not sure how to reproduce crated a token that timed out.

@larkox @hanzei is this related to this? https://github.com/mattermost/mattermost-plugin-zoom/issues/239

**Comment by hanzei** [comment]
@dipak-demansol Yes

**Comment by larkox** [comment]
I don't remember the timeline 100%, but in the old version, after you connect, if you wait for, I think, one hour, the token will be outdated and will force you to disconnect and connect again.

In order to test it, have two environments: one with the new version, and one with the old, and connect to zoom (on user level oauth) roughly at the same time. Then let that simmer for a few hours (or overnight), and try to create a new meeting. The old version should fail, saying that the token is expired, or something like that, and the new version should work without issues.

It would be useful also to have a third environment with account level app, to see if that is still working fine.

**Review by dipak-demansol — APPROVED** [review]
**LGTM**, Tested With Two server, 1st server had a market zoom plugin on a Chrome browser and 2nd server had a PR_code zoom app on the Brave browser, after 6 hour issue generated in the 1st server but the same issue not generated in the 2nd server.  
**Additional Info** :-  with the chrome browser and 2nd server when i create zoom meeting and join the meeting then i'm successfully join the meeting but in the brew browser with the 2nd server when i perform the same step then zoom don't identify me as a host and its wait for start the meeting. i will discuss with dylan about that.

**Pull request merged**